allennlp_models/vision/models/vilbert_image_retrieval.py

```diff
@@ -0,0 +1,163 @@
+        box_mask: torch.Tensor,
+        caption: TextFieldTensors,
+        label: torch.Tensor,
+    ) -> Dict[str, torch.Tensor]:
+        batch_size = box_features.shape[0]
+
+        if self.training:
+            # Shape: (batch_size, num_images, pooled_output_dim)
+            pooled_output = self.backbone(box_features, box_coordinates, box_mask, caption)[
+                "pooled_boxes_and_text"
+            ]
+
+            # Shape: (batch_size, num_images)
+            logits = self.classifier(pooled_output).squeeze(-1)
+            probs = torch.softmax(logits, dim=-1)
+
+            outputs = {"logits": logits, "probs": probs}
+            outputs = self._compute_loss_and_metrics(batch_size, outputs, label)
+
+            return outputs
+
+        else:
+            with torch.no_grad():
+                # Shape: (batch_size, num_images, pooled_output_dim)
+                self.backbone.eval()
+                pooled_output = self.backbone(box_features, box_coordinates, box_mask, caption)[
+                    "pooled_boxes_and_text"
+                ]
+                self.backbone.train()
+
+                # Shape: (batch_size, num_images)
+                scores = self.classifier(pooled_output).squeeze(-1)
+
+                # Shape: (batch_size, k)
+                rel_scores, indices = scores.topk(self.k, dim=-1)
+
+                # Shape: (batch_size)
+                pre_logits = torch.sum((indices == label.reshape(-1, 1)), dim=-1).float()
+
+                # Shape: (batch_size, 2)
+                # 1st column == 1 if we found the image in the top k, 2nd column == 1 if we didn't
+                logits = torch.stack((pre_logits, (pre_logits == 0).float()), dim=1)
+
+                outputs = {"logits": logits}
+                outputs = self._compute_loss_and_metrics(
+                    batch_size, outputs, torch.zeros(batch_size).long().to(logits.device)
+                )
+
+                return outputs
+
+    @overrides
+    def _compute_loss_and_metrics(
+        self,
+        batch_size: int,
+        outputs: torch.Tensor,
+        labels: torch.Tensor,
+    ):
+        outputs["loss"] = self.loss(outputs["logits"], labels)
+        self.accuracy(outputs["logits"], labels)
+        return outputs
+
+    @overrides
+    def get_metrics(self, reset: bool = False) -> Dict[str, float]:
+        return {
+            "accuracy": self.accuracy.get_metric(reset),
+        }
+
+    @overrides
+    def make_output_human_readable(
+        self, output_dict: Dict[str, torch.Tensor]
+    ) -> Dict[str, torch.Tensor]:
+        batch_tokens = []
+        for batch_index, batch in enumerate(output_dict["probs"]):
+            tokens = {}
+            for i, prob in enumerate(batch):
+                tokens[self.vocab.get_token_from_index(i, self.label_namespace)] = float(prob)
+            batch_tokens.append(tokens)
+        output_dict["tokens"] = batch_tokens
+        return output_dict
```

Comment:
This is copy&paste gore, isn't it? That looks like the function from VQA.

Reply:
Yup, I didn't see anything VQA-specific in it so I thought it looked reasonable, but let me know if I should change it.

Reply:
This function assembles `output_dict["tokens"]`, because it thinks the answers it produces are tokens. This won't work in this model at all.

We can have this function do nothing. I don't know exactly how we would represent "the right image" in a `Dict` in a human-readable way. I guess if we had the original filename, that would be neat to add, but it's a bit of work to plumb that all the way through.

Reply:
Oh I see, that makes sense. Thanks for clarifying that. I'll make it just return the output dictionary.